### PR TITLE
LoFTR inference basic PR [WIP]

### PR DESCRIPTION
**Pull request by ducha-aiki**
### Description
This is very basic adaptation of the LoFTR (https://github.com/zju3dv/LoFTR)  for kornia.
Things done: keep only inference part, remove dependency on einops and config library, download weights via torch.hub

Missing things:

- code restructuring to a single file
- using kornia operations whenever possible
- tests
- code style fix

### Status
**Work in progress*

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

File: kornia/feature/loftr/utils/geometry.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import torch
+
+
+@torch.no_grad()
+def warp_kpts(kpts0, depth0, depth1, T_0to1, K0, K1):
```
Comment (edgarriba) [inline kornia/feature/loftr/utils/geometry.py:5]: not now but this should go as core function in `kornia.geometry`. Add a todo or something to be aware about that

File: kornia/feature/loftr/loftr.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  'train_coarse_percent': 0.4,
+  'train_pad_num_gt_min': 200},
+ 'fine': {'d_model': 128,
+  'd_ffn': 128,
+  'nhead': 8,
+  'layer_names': ['self', 'cross'],
+  'attention': 'linear'}}
+
+
+class LoFTR(nn.Module):
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:54]: add docstring with reference and small snippet usage

Comment (edgarriba) [reply kornia/feature/loftr/loftr.py:54]: would be really helpful also to have description of the arguments in the configuration or make a selection with the more relevant ones.

Comment (ducha-aiki) [reply kornia/feature/loftr/loftr.py:54]: Do not review yet :) 
THe PR is far from ready, that is just a cropped version of original repo.

Comment (edgarriba) [reply kornia/feature/loftr/loftr.py:54]: ok - mark as draft then :)

Comment (ducha-aiki) [reply kornia/feature/loftr/loftr.py:54]: I thought, that [WIP] is actually the correct mark :)

File: kornia/feature/loftr/loftr.py [hunk 3]
```diff
@@ -9,54 +9,75 @@
...
+               'fine': {'d_model': 128,
+                        'd_ffn': 128,
+                        'nhead': 8,
+                        'layer_names': ['self', 'cross'],
+                        'attention': 'linear'}}
 
 
 class LoFTR(nn.Module):
-    def __init__(self, config=default_cfg, pretrained: Optional[str] = None):
+    r"""Module, which finds correspondences between two images.
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:54]:
```suggestion
    r"""Module, which finds correspondences between two images.
```

File: kornia/feature/loftr/loftr.py [hunk 4]
```diff
@@ -9,54 +9,75 @@
...
+    This is based on the original code from paper "LoFTR: Detector-Free Local
+    Feature Matching with Transformers". See :cite:`LoFTR2021` for more details.
+    If the distance matrix dm is not provided, :py:func:`torch.cdist` is used.
+
+    Args:
+        config: Dict with initiliazation paramers. Do not pass it, unless you know what you are doint`.
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:60]:
```suggestion
        config: Dict with initialization parameters. Do not pass it, unless you know what you are doing`.
```

File: kornia/feature/loftr/loftr.py [hunk 5]
```diff
@@ -9,54 +9,75 @@
...
+        pretrained: Download and set pretrained weights to the model. Options: 'outdoor', 'indoor'.
+                    'outdoor' is trained on the MegaDepth dataset and 'indoor'
+                    on the ScanNet.
+
+    Returns:
+        Dictionary with image correspondences and confidence scores
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:66]:
```suggestion
        Dictionary with image correspondences and confidence scores.
```

File: kornia/feature/loftr/loftr.py [hunk 6]
```diff
@@ -67,24 +88,33 @@ def __init__(self, config=default_cfg, pretrained: Optional[str] = None):
...
             if pretrained not in urls.keys():
                 raise ValueError(f"pretrained should be None or one of {urls.keys()}")
             pretrained_dict = torch.hub.load_state_dict_from_url(
-                 urls[pretrained], map_location=lambda storage, loc: storage)
+                urls[pretrained], map_location=lambda storage, loc: storage)
             self.load_state_dict(pretrained_dict['state_dict'])
+        self.eval()
 
-    def forward(self, data):
+    def forward(self, data: Dict) -> Dict:
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:100]:
```suggestion
    def forward(self, data: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
```

File: kornia/feature/loftr/loftr.py [hunk 7]
```diff
@@ -67,24 +88,33 @@ def __init__(self, config=default_cfg, pretrained: Optional[str] = None):
...
         """
-        Update:
+        Args:
             data (dict): {
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:103]:
```suggestion
            data: {
```
not sure how will render the one below

File: kornia/feature/loftr/loftr.py [hunk 8]
```diff
@@ -131,6 +161,15 @@ def forward(self, data):
         # 5. match fine-level
         self.fine_matching(feat_f0_unfold, feat_f1_unfold, data)
 
+        rename_keys = {"mkpts0_f": 'keypoints0',
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr.py:164]:
```suggestion
        rename_keys: Dict[str, str] = {"mkpts0_f": 'keypoints0',
```

File: kornia/feature/loftr/loftr_module/linear_attention.py [hunk 9]
```diff
@@ -16,7 +17,12 @@ def __init__(self, eps=1e-6):
         self.feature_map = elu_feature_map
         self.eps = eps
 
-    def forward(self, queries, keys, values, q_mask=None, kv_mask=None):
+    def forward(self,
+                queries: torch.Tensor,
+                keys: torch.Tensor,
+                values: torch.Tensor,
+                q_mask: Optional[torch.Tensor] = None,
+                kv_mask: Optional[torch.Tensor] = None):
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr_module/linear_attention.py:26]:
```suggestion
                kv_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
```

File: kornia/feature/loftr/loftr_module/transformer.py [hunk 10]
```diff
@@ -25,16 +27,20 @@ def __init__(self,
...
         # norm and dropout
         self.norm1 = nn.LayerNorm(d_model)
         self.norm2 = nn.LayerNorm(d_model)
 
-    def forward(self, x, source, x_mask=None, source_mask=None):
+    def forward(self,
+                x: torch.Tensor,
+                source: torch.Tensor,
+                x_mask: Optional[torch.Tensor] = None,
+                source_mask: Optional[torch.Tensor] = None):
```
Comment (edgarriba) [inline kornia/feature/loftr/loftr_module/transformer.py:42]:
```suggestion
                source_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
```

**Review by edgarriba — APPROVED** [review]

**Comment by ducha-aiki** [comment]
@edgarriba torch 1.6.0 tests are failing for LoFTR because 1.6.0 does not support directories in the pretrained models packaging. Can we bump our cuda tests to >1.6.0 pytorch version?

**Comment by edgarriba** [comment]
> @edgarriba torch 1.6.0 tests are failing for LoFTR because 1.6.0 does not support directories in the pretrained models packaging. Can we bump our cuda tests to >1.6.0 pytorch version?

ok, let me do in separate PR

File: test/feature/test_loftr.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+from torch.autograd import gradcheck
+
+import kornia.testing as utils  # test utils
+from kornia.feature import LoFTR
+from kornia.geometry import resize
+from kornia.testing import assert_close
+
+
+class TestLoFTR:
+    def test_pretrained_outdoor_smoke(self, device):
```
Comment (edgarriba) [inline test/feature/test_loftr.py:12]: @ducha-aiki ok, I guess here is where we should skip cuda version

Comment (ducha-aiki) [reply test/feature/test_loftr.py:12]: I haven't thought about such simple solution))) OK

Comment (edgarriba) [reply test/feature/test_loftr.py:12]: be aware of parsing issues with version as mentioned here too https://github.com/kornia/kornia/issues/1264
